### PR TITLE
[code-completion] Add type context for single-expression closures

### PR DESCRIPTION
**Pull request by benlangmuir**
When completing in the only expression of closure, use the return type of the closure as the type context for the code-completion. However, since code-completion may be on an incomplete input, we only use the return type to improve the quality of the result, not to mark it invalid, since (a) we may add another statement afterwards, or (b) if the context type is Void it doesn't need to match the value.

### CODE REVIEW

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@rintaro this is a bit different than we discussed.  Instead of handling Void specially, I keep track of whether the type context is from a single-expression body and in that case do not mark anything as type-relation "invalid".

**Comment by benlangmuir** [comment]
@swift-ci please test

File: lib/IDE/ExprContextAnalysis.cpp [hunk 1]
```diff
@@ -705,14 +713,28 @@ class ExprContextAnalyzer {
...
 
+  static bool isSingleExpressionBodyForCodeCompletion(BraceStmt *body) {
+    switch (body->getNumElements()) {
+    case 0:
+      // Nothing in the body except the code-completion token.
+      return true;
+    case 1:
+      return body->getElements()[0].is<Expr *>();
+    default:
+      return false;
```
Comment (rintaro) [inline lib/IDE/ExprContextAnalysis.cpp]: Do we already have test cases for this path? like:
```swift
var floatVal: Float = 2
let _: () -> Float = {
  #^COMPLETE^# // Should be normal completion.
  return 42
}
let _: () -> SomeEnum = {
  .#^COMPLETE^# // Should not provide anything.
  return 42
}
```
If not, please add them.

**Comment by benlangmuir** [comment]
@swift-ci please test

**Review by rintaro — COMMENTED** [review]
Implementation makes sense to me :)

File: lib/IDE/ExprContextAnalysis.cpp (hunk 1 above)
Comment (benlangmuir) [reply lib/IDE/ExprContextAnalysis.cpp]: Good catch, this code for handling `getNumElements() == 0` is from before I changed the parser to add the `CodeCompletionExpr` in `completeStmtOrExpr`.  I'll remove the 0 case and add the additional test.

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@swift-ci please test macOS

**Review by rintaro — APPROVED** [review]

**Pull request merged**